code/modules/smithing/anvil.dm

```diff
@@ -37,292 +8,242 @@
-		var/skillmod = user.mind.get_skill_level(/datum/skill/level/dwarfy/blacksmithing)/10 + 1
-		finalfailchance = max(0, finalfailchance / skillmod) //lv 2 gives 20% less to fail, 3 30%, etc
-	if((currentsteps > 10 || (rng && prob(finalfailchance))) && !artifact)
-		to_chat(user, "<span class='warning'>You overwork the metal, causing it to turn into useless slag!</span>")
-		var/turf/T = get_turf(user)
+	if(workpiece.height > 144 || workpiece.height < 0)
+		to_chat(user, "<span class='danger'> You ruin the [workpiece] by overworking it!</span>")
+		qdel(workpiece)
 		workpiece_state = FALSE
-		new /obj/item/stack/ore/slag(T)
-		currentquality = anvilquality
-		stepsdone = ""
-		currentsteps = 0
-		outrightfailchance = 1
-		artifactrolled = FALSE
-		if(user.mind.skill_holder)
-			user.mind.auto_gain_experience(/datum/skill/level/dwarfy/blacksmithing, 25, 400, silent = FALSE)
-	for(var/i in smithrecipes)
-		if(i == stepsdone)
-			var/turf/T = get_turf(user)
-			var/obj/item/smithing/create = smithrecipes[stepsdone]
-			var/obj/item/smithing/finisheditem = new create(T)
-			to_chat(user, "You finish your [finisheditem]!")
-			if(artifact)
-				to_chat(user, "It is an artifact, a creation whose legacy shall live on forevermore.") //todo: SSblackbox
-				currentquality = max(currentquality, 2)
-				finisheditem.quality = currentquality * 3//this is insane i know it's 1/2500 for most of the time and 0.8% at best
-				finisheditem.artifact = TRUE
-			else
-				finisheditem.quality = min(currentquality, itemqualitymax)
-			switch(finisheditem.quality)
-				if(-1000 to -8)
-					finisheditem.desc =  "It looks to be the most awfully made object you've ever seen."
-				if(-8)
-					finisheditem.desc =  "It looks to be the second most awfully made object you've ever seen."
-				if(-8 to 0)
-					finisheditem.desc =  "It looks to be barely passable as... whatever it's trying to pass for."
-				if(0)
-					finisheditem.desc =  "It looks to be totally average."
-				if(0 to INFINITY)
-					finisheditem.desc =  "It looks to be better than average."
-			workpiece_state = FALSE
-			finisheditem.set_custom_materials(workpiece_material)
-			currentquality = anvilquality
-			stepsdone = ""
-			currentsteps = 0
-			outrightfailchance = 1
-			artifactrolled = FALSE
-			if(user.mind.skill_holder)
-				user.mind.auto_gain_experience(/datum/skill/level/dwarfy/blacksmithing, 50, 10000000, silent = FALSE)
-			break
+		workpiece = null
+	var/datum/smith_recipe/recipe = workpiece.plan
+	var/list/pls = recipe.planlaststeps
+	for(var/i in pls)
+		to_chat(user, i)
+	if(!recipe)
+		to_chat(user, "No recipe")
+		return FALSE
+	if(workpiece.height >= recipe.target_height_min && workpiece.height <= recipe.target_height_max)
+		if(pls ~= workpiece.last3steps) //the fuck is 'quivelance'
+			generateitem(user)
+		else
+			to_chat(user, "WRONG3STEP")
+			return FALSE
+	else
+		return FALSE
+
+
+
+/obj/structure/anvil/proc/generateitem(mob/user)
+	to_chat(user, "generating item")
```

Comment:
Debug text. This could probably be made more descriptive to be more fitting for a player-facing message.